Authentication (with Redux Saga)

mzaintariq opened the pull request:
Added login/authentication using Spotify API.

mzaintariq committed:
Authentication / Login

mzaintariq committed:
Removed extra files/code

mzaintariq committed:
Dashboard

mzaintariq committed:
Added redux saga implementation to login/authentication

mzaintariq committed:
Added logo asset

mzaintariq committed:
Separating action files

mzaintariq committed:
scss

mzaintariq committed:
Formatting

mzaintariq committed:
Merge branch 'main' into redux-saga-auth

Fiju commented on src/App.js:
```diff
@@ -1,12 +1,26 @@
-import React from "react";
-import Header from "./components/header/Header";
-import "./App.css";
+import "./App.scss";
+import React, { useEffect } from "react";
+import Dashboard from "./components/dashboard/Dashboard";
+import Login from "./components/login/Login";
+import { useSelector, useDispatch } from "react-redux";
+import { getToken } from "./actions/index";
```
Sort the import statements as:
- Import from React library
- Import from third party libraries
- Import from project scope
- Assets

Fiju commented on src/store.js:
```diff
@@ -1,6 +1,13 @@
+
+const store = createStore(rootReducer, {}, applyMiddleware(...middleware));
+
+sagaMiddleware.run(watcherSaga);
```
`redux-immutable-state-invariant` consider adding this in you middlewares as well. It detect any mutation that is done by mistake in redux store

Fiju commented on src/sagas/rootSaga.js:
```diff
@@ -0,0 +1,7 @@
+import { GET_TOKEN } from "../actions";
+
+export function* watcherSaga() {
+  yield takeLatest(GET_TOKEN, handleGetToken);
```
Consider using `fork` side effect

Fiju commented on src/sagas/handlers/auth.js:
```diff
@@ -0,0 +1,12 @@
+  } catch (error) {
+    console.log(error);
+  }
+}
```
Let's get of handler directory and have module specific sagas in the top level of sagas directory

Fiju commented on src/reducers/index.js:
```diff
@@ -1,8 +1,8 @@
 
 const rootReducer = combineReducers({
-  check,
+  authReducer,
```
```suggestion
  auth: authReducer,
```

Fiju commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,22 @@
+const initialState = {
```
Convention to declare constant with capital snake case convention

Fiju commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,22 @@
+const authReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case "SET_TOKEN":
+      const data = action.payload;
```
```suggestion
      const { payload: data } = action;
```

Fiju commented on src/components/login/Login.scss:
```diff
@@ -0,0 +1,20 @@
+.login {
+  display: grid;
+  place-items: center;
+  min-height: 100vh;
+  background-color: black;
+}
+
+.login > img {
+  width: 80%;
+  max-width: 1000px;
+}
+
+.login > a {
+  padding: 20px;
+  border-radius: 30px;
+  background-color: #1db954;
+  font-weight: 800;
+  color: white;
+  text-decoration: none;
+}
```
With sass we can next our selectors within each other.
```suggestion
.login {
  display: grid;
  place-items: center;
  min-height: 100vh;
  background-color: black;
  
  & > img {
  width: 80%;
  max-width: 1000px;
  }
  
  & > a {
  padding: 20px;
  border-radius: 30px;
  background-color: #1db954;
  font-weight: 800;
  color: white;
  text-decoration: none;
}
}
```

Fiju commented on src/components/login/Login.js:
```diff
@@ -0,0 +1,15 @@
+import React from "react";
+import "./Login.scss";
```
```suggestion
import styles from './Login.module.scss';
```
Consider treating the assets as a module. It will prevent the style (class) leaks through the app.

Fiju commented on src/app/spotify.js:
```diff
@@ -0,0 +1,15 @@
+const authEndpoint = "https://accounts.spotify.com/authorize";
+const scopes = [
+  "user-read-currently-playing",
+  "user-read-recently-played",
+  "user-read-playback-state",
+  "user-top-read",
+  "user-modify-playback-state",
+];
+export const redirectUri = "http://localhost:3000/";
+export const clientId = "9c2b9616464949a9a3e68c21250663a0";
+export const clientSecret = "a866fad991494333ae0dcb8f7fef0fcf";
+export const loginUrl = `${authEndpoint}?client_id=${clientId}&redirect_uri=${redirectUri}&scope=${scopes.join(
+  "%20"
+)}&response_type=code&show_dialogue=true`;
+export const tokenEndpoint = "https://accounts.spotify.com/api/token";
```
Let's move all of this in an environment file. And declare everything in a new file, say `setting.js` after importing them

Fiju commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,15 @@
+export const GET_TOKEN = "GET_TOKEN";
```
Move all the types definitions in a new file `actionTypes.js`

Fiju commented on src/sagas/requests/auth.js:
```diff
@@ -0,0 +1,17 @@
+    },
+    body: formBody,
+  }).then((response) => response.json());
+}
```
Let's move this in a new file within service directory. Maybe have wrappers created to simplify the code.

mzaintariq committed:
Code review changes

- sort import statements
- add redux-immutable-state-invariant in middleware
- fork side effect in rootSaga
- module.scss
- environment variables in .env.local
- added api.js in services directory for all api requests
